cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java

```diff
@@ -36,6 +40,49 @@ public static MDSKey getMdsKey(NamespacedEntityId targetId, long time) {
     return builder.build();
   }
 
+  /**
+   * Extracts history time from the record
+   * @param rowKey row key from which time needs to be extracted
+   * @param type type of the entity
+   * @return timestamp in the rowKey
+   */
+  static long extractTime(byte[] rowKey, String type) {
+    MDSKey.Splitter keySplitter = new MDSKey(rowKey).split();
+    // Skip rowType
+    keySplitter.skipBytes();
+
+    // Skip targetId
+    if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ProgramId.class))) {
```

Comment:
If I understand correctly, then ```EntityIdKeyHelper.TYPE_MAP.get()``` returns a String representing a EntityTypeSimpleName. So you could write this as a 

```
switch(EntityTypeSimpleName.valueOf(type)) {
case: PROGRAM: 
  // call skipString 4 times; 
  break;
case APPLICATION:
case DATASET:
case STREAM:
  // skipString() 2x
  break;
...
}
```

Wouldn't that be simpler?